### PR TITLE
Detect duplicates when uploading an image

### PR DESCRIPTION
**Pull request by Tijani-Dia**
The goal of this PR is to detect duplicates when uploading an image.

For now, we just detect exact file duplicates. More sophisticated duplicates finding techniques are planned for the future.

When a duplicate is detected, a message is shown to indicate it (See images below).

**Upload single image view**

Screenshot from 2022-01-04 11-37-11

**Upload multiples images view**

Screenshot from 2022-01-04 11-05-00

* Do the tests still pass? Yes
* Does the code comply with the style guide? Yes
* For Python changes: Have you added tests to cover the new/fixed behaviour? Yes
* For front-end changes: Did you test on all of [Wagtail’s supported environments]? No
    * **Please list the exact browser and operating system versions you tested**. Firefox
    * **Please list which assistive technologies you tested**. None
* For new features: Has the documentation been updated accordingly? No

### CODE REVIEW

File: client/scss/components/_messages.status.scss
Comment (jacobtoppm) [inline client/scss/components/_messages.status.scss:7]: I think "duplicate" is too specific for a generic component here. Try `warning`, which mirrors what we use in the messages framework?

File: wagtail/images/static_src/wagtailimages/scss/add-multiple.scss
Comment (jacobtoppm) [inline wagtail/images/static_src/wagtailimages/scss/add-multiple.scss:128]: Why is this necessary? This feels like something that would be applicable to more than one message class.

File: wagtail/images/views/images.py
Comment (jacobtoppm) [inline wagtail/images/views/images.py:314]: Generally, try to lean away from concatenating translatable strings. Particularly since it has to start with a space, this could be confusing to translators, who might not get the associated context! It may be worth just repeating the start of the phrase, so a translator can adjust the whole "image added, but might be a duplicate" sentiment

File: wagtail/images/views/images.py
Comment (jacobtoppm) [inline wagtail/images/views/images.py:315]: Thinking as a user, if an image I've just uploaded is identified as a duplicate, my next question is: what's it a duplicate of? I think we should address this - as there might be multiple duplicates, though, this might be simplest to do by providing a redirect button to a listing view with the potential duplicates.

File: wagtail/images/views/multiple.py
Comment (jacobtoppm) [inline wagtail/images/views/multiple.py:45]: It's probably a little clearer to determine whether it's a duplicate here, rather than setting it on the view and then retrieving it.

File: wagtail/images/views/multiple.py
Comment (jacobtoppm) [inline wagtail/images/views/multiple.py:56]: If we're going to do this for all new images, might be worth us considering indexing `file_hash` for performance?

**Review by jacobtoppm — CHANGES_REQUESTED** [review]
Nice work! This will be a great addition. A few tweaks and things to think about - also, images can be uploaded via the modal as well, so we should probably cover that case as well.

**Comment by dest81** [comment]
Can we make this optional? I mean can we have something like `WAGTAILIMAGES_CHECK_DUPLICATES` in settings? For example in our site users can't see each other's images so we allow to upload the same image multiple times by different users. And I believe there are other scenarios when duplicates are possible.

**Comment by jacobtoppm** [comment]
@dest81 How are you not letting users see each others' images? If it's by lacking the "choose" permission for other collections, I think this PR (when complete) should only allow users to see duplicates from their set of choosable images. Potentially we could add an extra setting, but as we do have a huge number already and generally identifying duplicates seems like a huge positive, I'm reluctant to do that just by default.

**Comment by dest81** [comment]
> @dest81 How are you not letting users see each others' images? 

with `construct_image_chooser_queryset` hook

> If it's by lacking the "choose" permission for other collections, I think this PR (when complete) should only allow users to see duplicates from their set of choosable images.

this not in PR as I can see but even if it would be doesn't solve my issue

> Potentially we could add an extra setting, but as we do have a huge number already and generally identifying duplicates seems like a huge positive, I'm reluctant to do that just by default.

it is not bad thing to have many extra settings :), you still can do not allow duplicates by default


or maybe this:

```
    is_duplicate = (
           ImageModel.objects.filter(file_hash=image.file_hash).exists()
    )
```
could be done as model property which can be easily overridden in custom Image model. So if project has custom requirements/permissions developer can always fix it with minimal effort.

**Comment by jacobtoppm** [comment]
> > If it's by lacking the "choose" permission for other collections, I think this PR (when complete) should only allow users to see duplicates from their set of choosable images.
> 
> this not in PR as I can see but even if it would be doesn't solve my issue

It's not in the PR yet, no. It's something I've requested be added before we approve it.

> > Potentially we could add an extra setting, but as we do have a huge number already and generally identifying duplicates seems like a huge positive, I'm reluctant to do that just by default.
> 
> it is not bad thing to have many extra settings :), you still can do not allow duplicates by default

Hah - generally in the core team, we're trying not to default to adding new settings all the time. It does multiply the number of configurations we then have to officially support - and if the setting isn't likely to be useful to many people, the complication to benefit ratio gets a bit skewed. That said, I'd like us to find some way of supporting configurations like this, even if it does require a bit of extra code due to not being a very standard setup.

> or maybe this:
> 
> ```
>     is_duplicate = (
>            ImageModel.objects.filter(file_hash=image.file_hash).exists()
>     )
> ```
> 
> could be done as model property which can be easily overridden in custom Image model. So if project has custom requirements/permissions developer can always fix it with minimal effort.

That might be more viable, I think (though that code would always return True for a saved image as it would find its own hash present, and as I've discussed with @Tijani-Dia, we should be identifying the potential duplicates so we can display them to the user), or potentially as a method on a queryset or manager `ImageModel.objects.duplicates_of(image)` for example.

**Comment by zerolab** [comment]
@Tijani-Dia #8284 mentions it supersedes this PR. If that is the case, can we close this?

**Comment by Tijani-Dia** [comment]
Yes, we can close it now.

**Pull request closed without merging**
